### PR TITLE
Dummy under storage extension for tutorial

### PR DESCRIPTION
**Pull request by madanadit**
Dummy UFS which forwards everything to LocalUFS. 

Purpose: Use template for documenting a tutorial to develop UFS extensions

### CODE REVIEW

File: underfs/tutorial/pom.xml [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            </compilerArgs>
+            <testExcludes>
+              <!-- Skip UFS contract tests unless -DtestTutorialUFS="dummy://path" is specified -->
+              <exclude>**/DummyUnderFileSystemContractTest.java</exclude>
+            </testExcludes>
+          </configuration>
+        </plugin>
+        <plugin>
+          <groupId>org.codehaus.mojo</groupId>
+          <artifactId>findbugs-maven-plugin</artifactId>
```
Comment (aaudiber) [inline underfs/tutorial/pom.xml]: is it possible to remove findbugs?

File: underfs/tutorial/pom.xml [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+      </releases>
+      <snapshots>
+        <enabled>false</enabled>
+      </snapshots>
+    </repository>
+  </repositories>
+
+  <properties>
+    <alluxio.version>1.6.0-SNAPSHOT</alluxio.version>
+    <build.path>build</build.path>
```
Comment (aaudiber) [inline underfs/tutorial/pom.xml]: I don't think we ever use this now

File: underfs/tutorial/pom.xml (hunk 1 above)
Comment (madanadit) [reply underfs/tutorial/pom.xml]: removed, thanks

**Comment by madanadit** [comment]
@aaudiber PTAL. Thanks!

File: underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+import java.util.List;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Dummy {@link UnderFileSystem} implementation for tutorial.
+ */
+@ThreadSafe
+public class DummyUnderFileSystem extends BaseUnderFileSystem {
+  public static final String DUMMY_SCHEME = "dummy:/";
```
Comment (apc999) [inline underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java]: for `DUMMY_SCHEME`, do we require "dummy:/" or "dummy://"?

File: underfs/tutorial/README.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+```
+
+### Run Integration Tests
+
+```bash
+mvn test -DtestTutorialUFS=dummy://<folder>
+```
+
+### Mount
+
```
Comment (apc999) [inline underfs/tutorial/README.md]: don't we need to deploy this ufs jar first?

File: underfs/tutorial/README.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+Dummy Alluxio under storage implementation accessing local file system using the `dummy://` scheme.
+
+### Build
+
+```bash
+mvn package
+```
+
+### Run Integration Tests
+
```
Comment (apc999) [inline underfs/tutorial/README.md]: can you explain here by having `-DtestTutorialUFS=dummy://<folder>` we are running a contract test. without `-DtestTutorialUFS=dummy://<folder>`, only unit tests are triggered

File: underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java (hunk 3 above)
Comment (madanadit) [reply underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java]: we expect `bin/alluxio fs mount /mnt/dummy dummy://tmp` to write to `/tmp`. excluded the second `/` to not require `dummy:///tmp`

Comment (aaudiber) [reply underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java]: file URIs are [required by the spec](https://en.wikipedia.org/wiki/File_URI_scheme#Meaning_of_slash_character) to use `file:///path/`, so we should do the same here

Comment (madanadit) [reply underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java]: @apc999 @aaudiber done, thanks

File: underfs/tutorial/src/test/java/alluxio/underfs/dummy/DummyUnderFileSystemContractTest.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+import alluxio.underfs.AbstractUnderFileSystemContractTest;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.UnderFileSystemConfiguration;
+
+import com.google.common.base.Preconditions;
+import org.junit.BeforeClass;
+
+/**
+ * This UFS contract test {@link DummyUnderFileSystem} as the backing store.
```
Comment (aaudiber) [inline underfs/tutorial/src/test/java/alluxio/underfs/dummy/DummyUnderFileSystemContractTest.java]: fix grammar

File: underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystemFactory.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Factory for creating {@link DummyUnderFileSystem}.
+ */
+@ThreadSafe
+public class DummyUnderFileSystemFactory implements UnderFileSystemFactory {
+
+  /**
+   * Constructs a new {@link DummyUnderFileSystemFactory}.
+   */
+  public DummyUnderFileSystemFactory() {}
```
Comment (aaudiber) [inline underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystemFactory.java]: can omit this empty constructor

File: underfs/tutorial/pom.xml [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+      <name>Maven Repository</name>
+      <url>https://repo1.maven.org/maven2</url>
+      <releases>
+        <enabled>true</enabled>
+      </releases>
+      <snapshots>
+        <enabled>false</enabled>
+      </snapshots>
+    </repository>
+    <repository>
```
Comment (aaudiber) [inline underfs/tutorial/pom.xml]: I don't think we need the apache repository

File: underfs/tutorial/pom.xml [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        <enabled>false</enabled>
+      </snapshots>
+    </repository>
+  </repositories>
+
+  <properties>
+    <alluxio.version>1.6.0-SNAPSHOT</alluxio.version>
+    <java.version>1.7</java.version>
+    <junit.version>4.12</junit.version>
+    <log4j.version>1.2.17</log4j.version>
```
Comment (aaudiber) [inline underfs/tutorial/pom.xml]: this is unused

File: underfs/tutorial/pom.xml [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    <maven.version>3.3.9</maven.version>
+  </properties>
+
+  <dependencyManagement>
+    <dependencies>
+      <!-- External dependencies -->
+      <dependency>
```
Comment (aaudiber) [inline underfs/tutorial/pom.xml]: can this be removed?

File: underfs/tutorial/pom.xml [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+  <properties>
+    <alluxio.version>1.6.0-SNAPSHOT</alluxio.version>
+    <java.version>1.7</java.version>
+    <junit.version>4.12</junit.version>
+    <log4j.version>1.2.17</log4j.version>
+    <maven.version>3.3.9</maven.version>
+  </properties>
+
+  <dependencyManagement>
```
Comment (aaudiber) [inline underfs/tutorial/pom.xml]: dependency management seems like overkill for this project - could we include dependencies directly instead?

File: underfs/tutorial/pom.xml [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+          <execution>
+            <phase>package</phase>
+            <goals>
+              <goal>shade</goal>
+            </goals>
+            <configuration>
+              <filters>
+                <filter>
+                  <artifact>*:*</artifact>
+                  <excludes>
```
Comment (aaudiber) [inline underfs/tutorial/pom.xml]: are these excludes needed for this project? @apc999

File: underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Dummy {@link UnderFileSystem} implementation for tutorial.
```
Comment (aaudiber) [inline underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java]: This is a good place to mention that all operations are delegated to `LocalUnderFileSystem`

File: underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+ */
+@ThreadSafe
+public class DummyUnderFileSystem extends BaseUnderFileSystem {
```
Comment (aaudiber) [inline underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java]: The methods abstract methods in `BaseUnderFileSystem` are inconsistent. Sometimes they take options, sometimes they don't. Now that we are asking external contributors to extend this class, we should make it `@PublicApi` and consistent (every filesystem method should take options).

File: underfs/tutorial/pom.xml (hunk 11 above)
Comment (madanadit) [reply underfs/tutorial/pom.xml]: i added this on bin's suggestion. @apc999 what do you think?

File: underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java (hunk 14 above)
Comment (madanadit) [reply underfs/tutorial/src/main/java/alluxio/underfs/dummy/DummyUnderFileSystem.java]: let's address that in another PR

File: underfs/tutorial/pom.xml (hunk 12 above)
Comment (apc999) [reply underfs/tutorial/pom.xml]: I don't think they are required

File: underfs/tutorial/pom.xml (hunk 11 above)
Comment (apc999) [reply underfs/tutorial/pom.xml]: since this is a single module, I think dependency management is not necessary.

**Pull request merged**